feat(lint-style): enable running on downstream projects

grunweg opened the pull request:
Enable lint-style to run on downstream projects, by making the following modifications:
- allow passing an explicit list of libraries to lint: if nothing is passed, it lints `Mathlib`, `Archive` and `Counterexamples` (as before); otherwise, it lints precisely the passed modules
- only check init imports, undocumented scripts and the errors from `lint-style.py` when linting Mathlib
- make the style exceptions file configurable and optional: using the `nolints-file` flag, the exceptions file can be configured. If the flag is omitted, we try to find a file at `scripts/nolints-style.txt` --- and otherwise proceed with no style exceptions.
This means mathlib can continue unchanged, and downstream projects can either add an explicit exceptions file, or proceed without any exceptions.

After this PR, one should be able to run lint-style on a downstream project by `lake exe lint-style ProjectName`.

Prompted by [this zulip discussion](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/lint-style.20for.20downstream.20libraries).

---

- [x] depends on: #24570
- [x] depends on: #24953

(I did not test the last part.)

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
feat(lint-style): make the list of libraries configurable

No explicit input results in the current default value "Mathlib, Archive,
Counterexamples"; passing any library means just these are linted.

grunweg committed:
feat: only check init imports and undocumented scripts if we're linting mathlib

grunweg committed:
Update module docstring

adomani commented on scripts/lint-style.lean:
```diff
@@ -109,10 +111,16 @@ def lintStyleCli (args : Cli.Parsed) : IO UInt32 := do
-  for s in ["Archive.lean", "Counterexamples.lean", "Mathlib.lean"] do
-    allModuleNames := allModuleNames.append (← findImports s)
+  for s in libraries do
+    allModuleNames := allModuleNames.append (← findImports s!"{s}.lean")
```
I just noticed this: maybe using `System.FilePath.addExtension` is more robust?

grunweg replied:
Good idea, fixed.

grunweg committed:
Use .addExtension and file paths

joneugster reviewed (approved):
LGTM, thanks!

joneugster commented:
I've tested in a local project with mathlib set to the latest commit here and get

```
jon@computer TestNormal % lake exe lint-style TestNormal
uncaught exception: no such file or directory (error code: 2)
  file: scripts/nolints-style.txt
```

joneugster commented on scripts/lint-style.lean:
```diff
@@ -125,8 +134,10 @@ def lintStyleCli (args : Cli.Parsed) : IO UInt32 := do
   -- This script is re-run each time, hence is immune to such issues.
   let nolints ← IO.FS.lines ("scripts" / "nolints-style.txt")
```
I've tested in a local project with mathlib set to the latest commit here and get

```
jon@computer TestNormal % lake exe lint-style TestNormal
uncaught exception: no such file or directory (error code: 2)
  file: scripts/nolints-style.txt
```

grunweg replied:
Thanks for testing! I've pushed a fix.

grunweg replied:
Would you like to re-test?

grunweg committed:
feat: make the nolints file configurable

grunweg commented on scripts/lint-style.lean:
```diff
@@ -125,8 +134,10 @@ def lintStyleCli (args : Cli.Parsed) : IO UInt32 := do
   let mut numberErrors ← lintModules nolints allModuleNames style fix
-  if ← checkInitImports then numberErrors := numberErrors + 1
-  if !(← allScriptsDocumented) then numberErrors := numberErrors + 1
+  -- If we are linting mathlib, also check the init imports and for undocumented scripts.
```
I think the comment is fine as-is.

grunweg commented on scripts/lint-style.lean:
```diff
@@ -109,10 +111,17 @@ def lintStyleCli (args : Cli.Parsed) : IO UInt32 := do
-    allModuleNames := allModuleNames.append (← findImports s)
+  for s in libraries do
+    allModuleNames := allModuleNames.append
+      (← findImports <| (System.mkFilePath [s]).addExtension "lean")
```
I like the second one, and have used it.

grunweg committed:
chore: use eprintln to print error information

grunweg committed:
Review comment

joneugster commented:
Sorry for the late review! I retested and still didn't manaage to get this running.

Now the next error is:

```
error: `print-style-error.sh` exited with code 255
could not execute external process './scripts/print-style-errors.sh'
```

Moreover, while the added option `--nolints-file` is nice to have, it would be desirable that the existance of such an exception file would be optional (and not throw an error)

grunweg commented:
Thanks for testing. Right, `print-style-exceptions.sh` is another issue. That file is just for running `lint-style.py`; since the number of checks in there is really small now, I'm inclined to *not* invent another configuration option, but simply omit that check downstream (and work further to remove that file).

Making the exceptions file optional: good idea. I'll do so, when I return to this PR.

grunweg committed:
change: make nolints-style.py optional

If no such file exists, just proceed with no style exceptions.

grunweg committed:
chore: omit running lint-style.py when run outside of mathlib
That would require a print-style-exceptions.sh script in the right location:
let's not open that can of worms. Instead, the Python linters should be
rewritten in Lean, so that step can be removed.

grunweg commented:
I just addressed all the comments. I'm testing locally now.

grunweg committed:
Fixup

grunweg commented:
@joneugster Do you mind testing again, whenever you have a moment to spare on this PR? (Trying to create a test project hangs forever for me...) Thanks!

Vierkantor commented:
Looks like I just made a PR doing something similar: #24570 (thanks @grunweg for spotting this!) I think we should continue with this PR and close the other, since this is more featureful.

I would like to add the Lakefile-based autodetection of which modules to lint. That way, it should be very easy to set up style linters for downstream projects with no configuration needed at all.

Vierkantor commented:
Sorry for changing my mind. Having taken a closer look, there are multiple changes in this PR that can stand alone. One of them is not hardcoding which modules should be linted. And checking the lakefile is a more useful default than Mathlib, Archive and Counterexamples. So I believe that it would be more effective to do the module selection in #24570, independently of this one.

As for running certain linters depending on whether this is part of Mathlib or not, that is a good idea. I was thinking to do it a bit more generally using Lean options (in the same way that currently we have the syntax linters listening to `set_option linter.foo`). I have some ideas on how to implement this, hopefully I can come up with a PR soon.

A concern for the configurable nolints file location is that we might just be adding configurability for the sake of configurability: do you have a usecase where this would be helpful?

grunweg commented:
I broadly agree with your comment; adding auto-detection sounds nice.

My main motivation for configurability is to not impose a mathlib convention on the exceptions file location (for instance, inside a `scripts` directory: should we force downstream users to create one?). Carleson and sphere-eversion don't have a scripts directory, for instance (or wouldn't if not for `nolints.json`, which is also hard-coded).

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

joneugster commented:
I think it would be valuable to make this work! However, since there hasn't been any movement on this PR for about a year, I'm going to mark it as `please-adopt` and unassign myself. 

Please request my review or assign myself again should this PR be reviewed.
